daft/sql/sql_reader.py

```diff
@@ -0,0 +1,106 @@
+from __future__ import annotations
+
+import logging
+from urllib.parse import urlparse
+
+import pyarrow as pa
+
+logger = logging.getLogger(__name__)
+
+
+class SQLReader:
+    def __init__(
+        self,
+        sql: str,
+        url: str,
+        limit: int | None = None,
+        offset: int | None = None,
+        apply_limit_before_offset: bool | None = None,
+        projection: list[str] | None = None,
+        predicate: str | None = None,
+    ) -> None:
+        if limit is not None and offset is not None and apply_limit_before_offset is None:
+            raise ValueError("apply_limit_before_offset must be specified when limit and offset are both specified")
+
+        self.sql = sql
+        self.url = url
+        self.limit = limit
+        self.offset = offset
+        self.apply_limit_before_offset = apply_limit_before_offset
+        self.projection = projection
+        self.predicate = predicate
+
+    def read(self) -> pa.Table:
+        sql = self._construct_sql_query()
+        return self._execute_sql_query(sql)
+
+    def _construct_sql_query(self) -> str:
+        clauses = ["SELECT"]
+        if self.projection is not None:
+            clauses.append(", ".join(self.projection))
+        else:
+            clauses.append("*")
+
+        clauses.append(f"FROM ({self.sql}) AS subquery")
+
+        if self.predicate is not None:
+            clauses.append(f" WHERE {self.predicate}")
+
+        if self.limit is not None and self.offset is not None:
+            if self.apply_limit_before_offset is True:
+                clauses.append(f"ORDER BY 1 LIMIT {self.limit} OFFSET {self.offset}")
```

Comment:
Added order by 1 because Trino arbitrarily discards rows without a specified order by clause.  https://trino.io/docs/current/sql/select.html#offset-clause

Reply:
Actually, order by will introduce unnecessary sorting, so this might not be a good choice. I'll investigate a bit more on how the trino queries can be partitioned, worst case we don't do partitioned reads for Trino

Reply:
Actually this isn't just a Trino thing: https://www.postgresql.org/docs/current/queries-limit.html ```using different LIMIT/OFFSET values to select different subsets of a query result will give inconsistent results unless you enforce a predictable result ordering with ORDER BY. This is not a bug; it is an inherent consequence of the fact that SQL does not promise to deliver the results of a query in any particular order unless ORDER BY is used to constrain the order.``` 

so i think if we keep the limit and offsets we have no choice but to do order by, but then theres also potential issues if the order by column is not unique